File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/System/Runtime/InteropServices/JavaScript/SecondRuntimeTest.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System.Collections.Generic;
using System.IO;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Xunit;

namespace System.Runtime.InteropServices.JavaScript.Tests
{
    public partial class SecondRuntimeTest
    {
        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsBrowserDomSupportedOrNodeJS))]
        public static async Task RunSecondRuntimeAndTestStaticState()
        {
            var runId = Guid.NewGuid().ToString();
            await JSHost.ImportAsync("SecondRuntimeTest", "../SecondRuntimeTest.js?run=" + runId);

            Interop.State = 42;
            var state2 = await Interop.RunSecondRuntimeAndTestStaticState(runId);
            Assert.Equal(44, Interop.State);
            Assert.Equal(3, state2);
        }

        public static partial class Interop
        {
            public static int State { get; set; }

            [JSExport]
            public static int IncrementState() => ++State;

            [JSImport("runSecondRuntimeAndTestStaticState", "SecondRuntimeTest")]
            [return: JSMarshalAs<JSType.Promise<JSType.Number>>]
            internal static partial Task<int> RunSecondRuntimeAndTestStaticState(string guid);
        }
    }
}

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/System/Runtime/InteropServices/JavaScript/SecondRuntimeTest.js
export async function runSecondRuntimeAndTestStaticState(guid) {
    const { dotnet: dotnet2 } = await import('./_framework/dotnet.js?instance=2-' + guid);
    const runtime2 = await dotnet2
        .withConfig({
            forwardConsoleLogsToWS: false,
            diagnosticTracing: false,
            appendElementOnExit: false,
            logExitCode: false,
        })
        .create();

    const increment1 = await getIncrementStateFunction(App.runtime);
    const increment2 = await getIncrementStateFunction(runtime2);

    increment1();
    increment1();
    increment2();
    increment2();
    const state2 = increment2();
    return state2;
}

async function getIncrementStateFunction(runtime) {
    const exports = await runtime.getAssemblyExports("System.Runtime.InteropServices.JavaScript.Tests.dll");
    return exports.System.Runtime.InteropServices.JavaScript.Tests.SecondRuntimeTest.Interop.IncrementState;
}